### PR TITLE
Feature/ml streaming base

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
- Implement the base architecture of ML Streaming building blocks
#### What does this implement/fix? Explain your changes.

The base architecture of ML Streaming

### CODE REVIEW

**Comment by AHReccese** [comment]
@sepandhaghighi I will prepare and drop a video in 24 hours in order to get onboarded better

**Comment by AHReccese** [comment]
Oh god finally tests are executing successfully, I ran into some weird hard-to-debug issues, hopefully, they all are solved now.
Continue with your review :)

File: setup.py [hunk 1]
```diff
@@ -1,15 +1,30 @@
...
+    "fastapi",
+    "requests"
+]
+
+
+def get_extra_ml_streaming_requires():
+    """Extract ML streaming extra requirments."""
+    requirements = open("requirements.txt", "r").read()
+    all_reqs = list(filter(lambda x: x != "", requirements.split()))
+    return [req for req in all_reqs if any(streaming in req for streaming in STREAMING_REQUIREMENTS)]
```
Comment (sadrasabouri) [inline setup.py:20]: It's an over-kill, I guess. Let's use `string.startswith(.)` instead.

File: setup.py [hunk 2]
```diff
@@ -1,15 +1,30 @@
...
+
+
+def get_core_requires():
+    """Extract PyMilo core requirments."""
     requirements = open("requirements.txt", "r").read()
-    return list(filter(lambda x: x != "", requirements.split()))
+    all_reqs = list(filter(lambda x: x != "", requirements.split()))
+    return [req for req in all_reqs if not any(streaming in req for streaming in STREAMING_REQUIREMENTS)]
```
Comment (sadrasabouri) [inline setup.py:27]: again here.

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,6 @@
 numpy>=1.9.0
 scikit-learn>=0.22.2
 scipy>=0.19.1
+uvicorn>=0.14.0
+fastapi>=0.68.0
+requests>=2.0.0
```
Comment (sadrasabouri) [inline requirements.txt:6]: I was thinking maybe we can add this file to another requirements file.
What do you think @sepandhaghighi ?

File: tests/test_ml_streaming/scenarios/scenario2.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    # 1.
+    x_train, y_train, x_test, y_test = prepare_simple_regression_datasets()
+    linear_regression = LinearRegression()
+    client = PymiloClient(model=linear_regression, mode="LOCAL")
+
+    # 2.
+    client.upload()
+
+    # 3.
+    client.toggle_mode()
```
Comment (sadrasabouri) [inline tests/test_ml_streaming/scenarios/scenario2.py:26]: I don't like the `toggle_mode` design. Finding out the current `mode` of a client connection, user should go through the code and count the number of toggles.

File: pymilo/streaming/pymilo_server.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            'args': request.args,
+            'kwargs': request.kwargs
+        }
+        args = gdst.deserialize(arguments, 'args', None)
+        kwargs = gdst.deserialize(arguments, 'kwargs', None)
+        output = retrieved_attribute(*args, **kwargs)
+        if isinstance(output, type(self._model)):
+            self._model = output
+            return "The ML model has been updated in place."
+        return gdst.serialize({'output': output}, 'output', None)
```
Comment (sadrasabouri) [inline pymilo/streaming/pymilo_server.py:39]: This one should be fixed later. It's better not to have multiple types of outputs.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]
Great job Amirhossein,
Take a look at my comments.

File: setup.py (hunk 1 above)
Comment (AHReccese) [reply setup.py:20]: What do you mean by `string.startswith(.)`?

File: requirements.txt (hunk 3 above)
Comment (AHReccese) [reply requirements.txt:6]: hmmm, I agree, I will do that.

File: tests/test_ml_streaming/scenarios/scenario2.py (hunk 4 above)
Comment (AHReccese) [reply tests/test_ml_streaming/scenarios/scenario2.py:26]: I agree, please drop your suggestion. I will think about it too.

File: pymilo/streaming/pymilo_server.py (hunk 5 above)
Comment (AHReccese) [reply pymilo/streaming/pymilo_server.py:39]: I can change it with `None`, agree?

File: setup.py (hunk 1 above)
Comment (AHReccese) [reply setup.py]: recheck, I've rewritten this part.

File: setup.py (hunk 2 above)
Comment (AHReccese) [reply setup.py]: recheck, I've rewritten this part.

File: requirements.txt (hunk 3 above)
Comment (AHReccese) [reply requirements.txt]: applied.

File: pymilo/streaming/pymilo_server.py (hunk 5 above)
Comment (AHReccese) [reply pymilo/streaming/pymilo_server.py]: Applied.

File: setup.py [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
 except ImportError:
     from distutils.core import setup
 
 
-def get_requires():
+def get_requires(streaming=False):
     """Read requirements.txt."""
-    requirements = open("requirements.txt", "r").read()
+    reqs = "requirements.txt" if not streaming else "streaming-requirements.txt"
+    requirements = open(reqs, "r").read()
```
Comment (sadrasabouri) [inline setup.py:12]: This looks like a hot fix, how about a dictionary that maps `streaming` and `core` to `streaming-requirements.txt` and `requirements.txt`.

File: tests/test_ml_streaming/scenarios/scenario2.py (hunk 4 above)
Comment (sadrasabouri) [reply tests/test_ml_streaming/scenarios/scenario2.py:26]: having a `mode` argument which sets to `local` as default.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]
That's much better now, thanks.
Minor comments.

File: pymilo/streaming/pymilo_server.py (hunk 5 above)
Comment (sadrasabouri) [reply pymilo/streaming/pymilo_server.py]: I meant the whole function, but I guess that's fine for now. You can ignore this comment.

File: tests/test_ml_streaming/scenarios/scenario2.py (hunk 4 above)
Comment (AHReccese) [reply tests/test_ml_streaming/scenarios/scenario2.py:26]: hmmm, so your suggestion is that we keep this syntax and add an argument named mode?
like this:
`client.toggle_mode(mode="DELEGATE")`

Comment (AHReccese) [reply tests/test_ml_streaming/scenarios/scenario2.py:26]: FYI:
we do have an attribute named `_mode` in the `PyMiloClient` too.

File: pymilo/streaming/pymilo_server.py (hunk 5 above)
Comment (AHReccese) [reply pymilo/streaming/pymilo_server.py]: Sorry, I forgot to replace the "Error" part, It is fixed now.

File: setup.py (hunk 6 above)
Comment (AHReccese) [reply setup.py:12]: I know your taste :), sure I will do.

Comment (AHReccese) [reply setup.py]: applied

File: pymilo/streaming/pymilo_server.py (hunk 5 above)
Comment (AHReccese) [reply pymilo/streaming/pymilo_server.py]: applied

File: tests/test_ml_streaming/scenarios/scenario2.py (hunk 4 above)
Comment (sadrasabouri) [reply tests/test_ml_streaming/scenarios/scenario2.py]: No I meant `client.predict(..., mode='LOCAL')`

**Review by sadrasabouri — APPROVED** [review]
LGTM, regarding `toggle_mode`, let's have @sepandhaghighi's comment too.

**Review by sepandhaghighi — APPROVED** [review]
@AHReccese Thanks for your efforts 🔥 

This pull request is a base PR and looks good to me, but I have some general comments that should be addressed in the next PRs:

1. I believe that we should add `pydantic` to both `streaming-requirements.txt` and `dev-requirements.txt`.

2. We should separate `streaming` tests from core tests(Test the core features of PyMilo without installing `streaming` dependencies, and then install and test the streaming features.)

3. Update all docstrings

4. I suggest using `enum` instead of `string` for parameters like `mode` (for example `ClientMode.Local` instead of `LOCAL`).

**Comment by AHReccese** [comment]
OK, sure.
I will address them all in the next PRs.
@sepandhaghighi

**Comment by AHReccese** [comment]
> @AHReccese Thanks for your efforts 🔥
> 
> This pull request is a base PR and looks good to me, but I have some comments that should be addressed in the next PRs:
> 
> 1. I believe that we should add `pydantic` to both `streaming-requirements.txt` and `dev-requirements.txt`.
> 2. We should separate `streaming` tests from core tests(Test the core features of PyMilo without installing `streaming` dependencies, and then install and test the streaming features.)
> 3. Update all docstrings
> 4. I suggest using `enum` instead of `string` for parameters like `mode` (for example `ClientMode.Local` instead of `LOCAL`).

1,3,4 are resolved.

**Pull request merged**